core/common/src/main/java/alluxio/collections/LockCache.java

```diff
@@ -0,0 +1,201 @@
+  static final float SOFT_LIMIT_RATIO = 0.9f;
+
+  private final Map<K, ValNode<V>> mCache;
+
+  /* A suggested maximum size for the cache */
+  private final int mHardLimit;
+  private final int mSoftLimit;
+  private final Function<? super K, ? extends V> mDefaultLoader;
+  private Iterator<Map.Entry<K, ValNode<V>>> mIterator;
+  private final Lock mEvictLock;
+
+  /**
+   * Constructor for a lock cache.
+   *
+   * @param defaultLoader specify a function to generate a value based on a key
+   * @param initialSize initial size of the cache
+   * @param maxSize maximum size of the cache
+   * @param concurrencyLevel concurrency level of the cache
+   */
+  public LockCache(@Nullable Function<? super K, ? extends V> defaultLoader, int initialSize,
+      int maxSize, int concurrencyLevel) {
+    mDefaultLoader = defaultLoader;
+
+    mHardLimit = maxSize;
+    mSoftLimit = (int) Math.round(SOFT_LIMIT_RATIO * maxSize);
+    mCache = new ConcurrentHashMap<>(initialSize, DEFAULT_LOAD_FACTOR, concurrencyLevel);
+    mIterator = mCache.entrySet().iterator();
+    mEvictLock = new ReentrantLock();
+  }
+
+  private void evictIfOverLimit() {
+    long numToEvict = mCache.size() - mSoftLimit;
+
+    if (numToEvict <= 0) {
+      return;
+    }
+    // this will block if every lock has a reference on them.
+    if (mEvictLock.tryLock()) {
+      try {
+        // This thread is the evictor
+        while (numToEvict > 0) {
+          if (!mIterator.hasNext()) {
+            mIterator = mCache.entrySet().iterator();
+          }
+          Map.Entry<K, ValNode<V>> candidateMapEntry = mIterator.next();
+          ValNode<V> candidate = candidateMapEntry.getValue();
+          if (candidate.mIsAccessed) {
+            candidate.mIsAccessed = false;
+          } else {
+            if ((!candidate.mIsNew) && candidate.mRefCount.compareAndSet(0, Integer.MIN_VALUE)) {
+              // the value object can be evicted, at the same time we make refCount minValue
+              mIterator.remove();
+              numToEvict--;
+            }
+          }
+        }
+      } finally {
+        mEvictLock.unlock();
+      }
+    }
+  }
+
+  /**
+   * get the value from the cache.
+   *
+   * @param key the key to look up the cache
+   * @return the value contained in the cache, if it is already in cache,
+   * otherwise generate an entry based on the loader
+   */
+  public ValNode<V> get(final K key) {
+    Preconditions.checkNotNull(key, "key can not be null");
+    ValNode<V> oldCacheEntry = null;
+    ValNode<V> cacheEntry;
+    while (true) {
+      // repeat until we get a cacheEntry that is not in the process of being removed.
+      cacheEntry = mCache.computeIfAbsent(key, k -> {
+        if (mCache.size() >= mHardLimit) {
+          return null;
+        } else {
+          return new ValNode<>(mDefaultLoader.apply(k));
+        }
+      });
+
+      if (cacheEntry == null) {
+        // cache is at hard limit
```

Comment:
I feel we might get a lot of log messages if we log here..

Reply:
we could do something like keep track of the last time we logged a warning here, and only log every 5 minutes or so

just worried that it will be hard to know without a log warning if this is affecting performance

Reply:
logging every minute now